src/V3List.h

```diff
@@ -140,4 +141,199 @@ class V3ListEnt final {
 
+// The list links, to be added as a member to 'Element'. They are considered mutable, even in a
+// 'const Element', as they are only used for storing 'Element' in a collection.
+template <typename Base>
+class V3List2Links final {
+    // The List itself, but nothing else can access the link pointers
+    template <typename B, V3List2Links<B> B::*P, typename>
+    friend class V3List2;
+
+    Base* m_nextp = nullptr;  // Next element in list
+    Base* m_prevp = nullptr;  // Previous element in list
+
+public:
+    V3List2Links() = default;
+    ~V3List2Links() {
+#ifdef VL_DEBUG
+        m_nextp = reinterpret_cast<Base*>(1);
+        m_prevp = reinterpret_cast<Base*>(1);
+#endif
+    }
+    VL_UNCOPYABLE(V3List2Links);
+    VL_UNMOVABLE(V3List2Links);
+};
+
+template <typename Base, V3List2Links<Base> Base::*LinksPointer, typename Element = Base>
+class V3List2 final {
+    static_assert(std::is_base_of<Base, Element>::value, "'Element' must be a subtype of 'Base");
+    Base* m_headp = nullptr;
+    Base* m_tailp = nullptr;
+
+    // Given the Element, return the Links. The links are always mutable, even in const elements.
+    VL_ATTR_ALWINLINE
+    static V3List2Links<Base>& toLinks(const Base& element) {
+        return const_cast<Base&>(element).*LinksPointer;
+    }
+
+    // Bare-bones iterator class for List. This is just enough to support range based for loops and
+    // basic usage. Feel free to extend as required.
+    template <typename IteratorElement>
+    class Iterator final {
+        static_assert(std::is_same<IteratorElement, Element>::value
+                          || std::is_same<IteratorElement, const Element>::value,
+                      "'Iterator' must be used with element type only");
+
+        // The List itself, but nothing else can construct iterators
+        template <typename B, V3List2Links<B> B::*P, typename>
+        friend class V3List2;
+
+        using SelfType = Iterator<IteratorElement>;
+
+        Base* m_currp;  // Currently iterated element, or 'nullptr' for 'end()'
+
+        Iterator(Base* elementp)
+            : m_currp{elementp} {
+            VL_PREFETCH_RW(elementp);
+        }
+        Iterator(std::nullptr_t)
+            : m_currp{nullptr} {}
+
+    public:
+        Iterator(const SelfType& other)
+            : m_currp{other.m_currp} {}
+        SelfType& operator=(const SelfType& other) {
+            m_currp = other.m_iterator;
+            return *this;
+        }
+        VL_UNMOVABLE(Iterator);
+
+        // Dereference
+        IteratorElement& operator*() const {
+            UASSERT(m_currp, "Dereferencing end of list iterator");
+            return *static_cast<IteratorElement*>(m_currp);
+        }
+        // Pre increment
+        SelfType& operator++() {
+            UASSERT(m_currp, "Pre-incrementing end of list iterator");
+            m_currp = toLinks(*m_currp).m_nextp;
+            if (VL_LIKELY(m_currp)) VL_PREFETCH_RW(m_currp);
```

Comment:
The iterators can do automatic prefetching for you.

Reply:
This might not be a gain, there's a cost in the hard to predict branch (and can't remove the branch due to perf issues on some Intel CPUs), please benchmark it specifically.

Reply:
That prefetch is significant in DFG usage, but I check the performance.

Also, the condition should almost always be true, it's only false when you step onto the end of the list, which in not often unless you have tiny lists (generally we don't), so I thought it would be pretty easy to predict.